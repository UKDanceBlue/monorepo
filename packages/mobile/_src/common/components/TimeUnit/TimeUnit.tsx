import { Text, View } from "native-base";
import { useWindowDimensions } from "react-native";

import { useThemeColors, useThemeFonts } from "../../customHooks";

const validUnits = ["sec", "min", "hours", "days", "months", "years"] as const;

/**
 * A label for a unit of time
 * @param props The react props
 * @param props.value The value of time
 * @param props.unit The unit of time (plural)
 */
const TimeUnit = ({
  value,
  unit,
}: {
  value: number | undefined;
  unit: (typeof validUnits)[number];
}) => {
  const { headingBold, body } = useThemeFonts();
  const { fontScale } = useWindowDimensions();
  const { primary } = useThemeColors();

  if (!validUnits.includes(unit)) {
    throw new Error(`Invalid unit: ${unit}`);
  }

  return (
    <View
      bgColor={`${primary[700]}BD`}
      marginX={2}
      style={{ alignItems: "center", paddingLeft: 7, paddingRight: 7, flex: 1 }}
      minWidth={30 * fontScale}
    >
      <Text
        color="secondary.400"
        fontSize="3xl"
        fontFamily={headingBold}
        flex={1}
        // minHeight={fontScale * 40} This seems to break things at least on android
      >
        {Math.max(0, value ?? 0)}
      </Text>
      <Text
        color="secondary.400"
        fontSize="xl"
        fontFamily={body}
        flex={1}
        minHeight={fontScale * 30}
      >
        {unit}
      </Text>
    </View>
  );
};

export default TimeUnit;
