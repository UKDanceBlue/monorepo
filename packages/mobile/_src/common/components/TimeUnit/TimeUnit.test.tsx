import { describe, expect, it, vitest } from "vitest";

import { renderWithNativeBase } from "../../../../test-helpers/NativeBase";
import TimeUnit from ".";

describe("<TimeUnit />", () => {
  it("renders 0 with a negative value", () => {
    const tree = renderWithNativeBase(<TimeUnit value={-1} unit={"sec"} />);

    const valueElement = tree.queryAllByText("0");
    expect(valueElement).toHaveLength(1);

    const unitElement = tree.queryAllByText("sec");
    expect(unitElement).toHaveLength(1);
  });

  it("renders 0 with a nullish value", () => {
    // @ts-expect-error Testing nullish values
    const tree = renderWithNativeBase(<TimeUnit value={null} unit={"sec"} />);

    const valueElement = tree.queryAllByText("0");
    expect(valueElement).toHaveLength(1);

    const unitElement = tree.queryAllByText("sec");
    expect(unitElement).toHaveLength(1);
  });

  it("throws when passed an invalid type", () => {
    const mockedConsoleError = vitest
      .spyOn(console, "error")
      .mockImplementation(() => undefined);

    const invalidUnit = "an invalid unit";
    expect(() =>
      // @ts-expect-error Testing an invalid unit
      renderWithNativeBase(<TimeUnit value={1} unit={invalidUnit} />)
    ).toThrow(`Invalid unit: ${invalidUnit}`);

    mockedConsoleError.mockRestore();
  });

  it("renders correctly with a singular unit", () => {
    const tree = renderWithNativeBase(
      <TimeUnit value={1} unit={"sec"} />
    ).toJSON() as unknown;

    expect(tree).toMatchSnapshot();
  });
  it("renders correctly with a plural unit", () => {
    const tree = renderWithNativeBase(
      <TimeUnit value={10} unit={"sec"} />
    ).toJSON() as unknown;

    expect(tree).toMatchSnapshot();
  });
});
