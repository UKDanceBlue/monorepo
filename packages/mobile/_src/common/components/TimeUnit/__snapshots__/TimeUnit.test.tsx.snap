// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimeUnit /> renders correctly with a plural unit 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={{}}
    style={
      [
        {
          "backgroundColor": "#0e7490BD",
          "marginLeft": 8,
          "marginRight": 8,
          "minWidth": 60,
        },
        {
          "alignItems": "center",
          "flex": 1,
          "paddingLeft": 7,
          "paddingRight": 7,
        },
      ]
    }
  >
    <Text
      dataSet={{}}
      style={
        {
          "backgroundColor": undefined,
          "color": "#f472b6",
          "flex": 1,
          "fontFamily": undefined,
          "fontSize": 30,
          "fontStyle": "normal",
          "fontWeight": "400",
          "letterSpacing": 0,
          "lineHeight": 45,
          "minHeight": 320,
          "textDecorationLine": undefined,
        }
      }
    >
      10
    </Text>
    <Text
      dataSet={{}}
      style={
        {
          "backgroundColor": undefined,
          "color": "#f472b6",
          "flex": 1,
          "fontFamily": undefined,
          "fontSize": 20,
          "fontStyle": "normal",
          "fontWeight": "400",
          "letterSpacing": 0,
          "lineHeight": 30,
          "minHeight": 60,
          "textDecorationLine": undefined,
        }
      }
    >
      sec
    </Text>
  </View>
</RNCSafeAreaProvider>
`;

exports[`<TimeUnit /> renders correctly with a singular unit 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={{}}
    style={
      [
        {
          "backgroundColor": "#0e7490BD",
          "marginLeft": 8,
          "marginRight": 8,
          "minWidth": 60,
        },
        {
          "alignItems": "center",
          "flex": 1,
          "paddingLeft": 7,
          "paddingRight": 7,
        },
      ]
    }
  >
    <Text
      dataSet={{}}
      style={
        {
          "backgroundColor": undefined,
          "color": "#f472b6",
          "flex": 1,
          "fontFamily": undefined,
          "fontSize": 30,
          "fontStyle": "normal",
          "fontWeight": "400",
          "letterSpacing": 0,
          "lineHeight": 45,
          "minHeight": 320,
          "textDecorationLine": undefined,
        }
      }
    >
      1
    </Text>
    <Text
      dataSet={{}}
      style={
        {
          "backgroundColor": undefined,
          "color": "#f472b6",
          "flex": 1,
          "fontFamily": undefined,
          "fontSize": 20,
          "fontStyle": "normal",
          "fontWeight": "400",
          "letterSpacing": 0,
          "lineHeight": 30,
          "minHeight": 60,
          "textDecorationLine": undefined,
        }
      }
    >
      se
    </Text>
  </View>
</RNCSafeAreaProvider>
`;
